rfd/0124-plugin-secrets.md

```diff
@@ -0,0 +1,174 @@
+---
+authors: Michael Wilson (michael.wilson@goteleport.com)
+state: draft
+---
+
+# RFD 124 - Plugin static credentials
+
+### Required Approvers
+
+* Engineering @r0mant, @justinas
+* Security @reed
+* Product: @klizhentas
+
+## What
+
+Allow storage of plugin static credentials within Teleport in a safe and secure way.
+
+## Why
+
+As we grow our plugin support, we have discovered that the various services we
+hope to integrate do not offer a uniform mechanism for configuration. For instance,
+Slack uses oauth2, but Okta requires an API token, and Jamf requires a username and
+password.
+
+Today, the plugin houses a credentials object that allows for this sort of storage.
+However, unlike an oauth authorization code, static credentials pose a much greater risk of
+exposure, and rotating credentials can be a time consuming process.  To reduce the risks associated
+with storing static credentials within Teleport and to allow for easier credentials rotation,
+we would like to create a new resource within Teleport that is handled in an explicitly secure
+way.
+
+## Details
+
+### The `PluginStaticCredentials` object
+
+A new object will be needed to house static credentials. At present it will support:
+
+- API tokens
+- Usernames and passwords
+- OAuth client ID and secret
+
+This object has very strict access control requirements that will be explained in detail
+in the **Access Control** section. On creation, the plugin and static credentials will
+receive a UUID called the `plugin-label`. The static credentials will have an internal
+label that receives this value. Additionally, one plugin may have _many_ credentials
+that are disambiguated with an additional label `plugin-credentials-type`. This is an
+arbitrary plugin specific string that will allow plugin authors to refer to multiple
+static
+credentials if needed.
+
+```mermaid
+flowchart LR
+  Plugin
+  Okta[Okta Credentials]
+  Slack[Slack Credentials]
+  OpsGenie[OpsGenie Credentials]
+
+  Plugin-->|credentials type okta|Okta
+  Plugin-->|credentials type slack|Slack
+  Plugin-->|credentials type opsgenie|OpsGenie
+```
+
+```yaml
+kind: plugin_static_credentials
+version: v1
+metadata:
+  name: credentials-name
+  labels:
+    teleport.internal/plugin-label: 48398071-9860-4be6-9f29-ca6ca5bc7155 # unique plugin label.
```

Comment:
Where is this unique label coming from? Is it always required or would we support free-form label matching like for other resources (see my example selector on the plugin resource in one of previous comments).

Reply:
This is something that would be generated upon plugin creation. This would be used to ensure that only plugin A can read credentials intended for plugin A, if that makes sense.